src/System.Net.Security/src/System/Net/_NativeSSPI.cs

```diff
@@ -0,0 +1,400 @@
+        {
+            return SafeDeleteContext.InitializeSecurityContext(ref credential, ref context, targetName, inFlags, endianness, null, inputBuffers, outputBuffer, ref outFlags);
+        }
+
+
+        public int EncryptMessage(SafeDeleteContext context, Interop.Secur32.SecurityBufferDescriptor inputOutput, uint sequenceNumber)
+        {
+            int status = (int)Interop.SecurityStatus.InvalidHandle;
+            bool ignore = false;
+
+            context.DangerousAddRef(ref ignore);
+            status = Interop.Secur32.EncryptMessage(ref context._handle, 0, inputOutput, sequenceNumber);
+            context.DangerousRelease();
+
+            return status;
+        }
+
+        public unsafe int DecryptMessage(SafeDeleteContext context, Interop.Secur32.SecurityBufferDescriptor inputOutput, uint sequenceNumber)
+        {
+            int status = (int)Interop.SecurityStatus.InvalidHandle;
+            uint qop = 0;
+
+            try
+            {
+                bool ignore = false;
+                context.DangerousAddRef(ref ignore);
+                status = Interop.Secur32.DecryptMessage(ref context._handle, inputOutput, sequenceNumber, &qop);
+            }
+            finally
+            {
+                context.DangerousRelease();
+            }
+
+            const uint SECQOP_WRAP_NO_ENCRYPT = 0x80000001;
+            if (status == 0 && qop == SECQOP_WRAP_NO_ENCRYPT)
+            {
+                GlobalLog.Assert("Secur32.DecryptMessage", "Expected qop = 0, returned value = " + qop.ToString("x", CultureInfo.InvariantCulture));
+                throw new InvalidOperationException(SR.net_auth_message_not_encrypted);
+            }
+
+
+            return status;
+        }
+
+        public int MakeSignature(SafeDeleteContext context, Interop.Secur32.SecurityBufferDescriptor inputOutput, uint sequenceNumber)
+        {
+            int status = (int)Interop.SecurityStatus.InvalidHandle;
+            bool ignore = false;
+
+            context.DangerousAddRef(ref ignore);
+            const uint SECQOP_WRAP_NO_ENCRYPT = 0x80000001;
+            status = Interop.Secur32.EncryptMessage(ref context._handle, SECQOP_WRAP_NO_ENCRYPT, inputOutput, sequenceNumber);
+            context.DangerousRelease();
+            return status;
+        }
+
+        public unsafe int VerifySignature(SafeDeleteContext context, Interop.Secur32.SecurityBufferDescriptor inputOutput, uint sequenceNumber)
+        {
+            int status = (int)Interop.SecurityStatus.InvalidHandle;
+
+            uint qop = 0;
+            try
+            {
+                bool ignore = false;
+                context.DangerousAddRef(ref ignore);
+                status = Interop.Secur32.DecryptMessage(ref context._handle, inputOutput, sequenceNumber, &qop);
```

Comment:
We need to add a Unit/Functional test to understand the implications. Tracking with #3114.